lambda/varianceUtils.js

```diff
@@ -0,0 +1,40 @@
+function calculateVariance(
+  firstYearValue = null,
+  secondYearValue = null,
+  thirdYearValue = null,
+  currentValue,
+  variancePercentage
+) {
+  let numberOfYearsProvided = 0;
+  numberOfYearsProvided += firstYearValue ? 1 : 0;
+  numberOfYearsProvided += secondYearValue ? 1 : 0;
+  numberOfYearsProvided += thirdYearValue ? 1 : 0;
+
+  const averageHistoricValue =
+    ((firstYearValue ? firstYearValue : 0) +
+      (secondYearValue ? secondYearValue : 0) +
+      (thirdYearValue ? thirdYearValue : 0)) /
+    numberOfYearsProvided;
+
+  const percentageChange =
+    (currentValue - averageHistoricValue) / averageHistoricValue;
+  const percentageChangeAbs = Math.abs(percentageChange);
+
+  let varianceErrorMessage = `Variance triggered: ${
+    percentageChangeAbs >= 0 ? "+" : "-"
+  }${String(percentageChangeAbs * 100)}%`;
```

Comment:
This creates rounding errors.  For example, if you use the following test case:

`const res = calculateVariance(8.5, 8.5, 8.5, 10.8, 0.2);`

It creates `percentageChange` = `0.27058823529411774` which turns into `0.27058823529411775` when you multiply by 10 on this line.  We may want to trim this to just 2 digits of precision `parseFloat(percentageChange).toPrecision(2)*100` or the like.

Incorrect calculation of `averageHistoricValue`: The calculation of `averageHistoricValue` is incorrect because it divides the sum of historical values by `numberOfYearsProvided`, which may be zero if no historical values are provided. This will result in a division by zero error. Additionally, if only one or two historical values are provided, the average will not be calculated correctly. The calculation should divide the sum by the total number of historical values, not just the number of years provided.

Inconsistent handling of optional parameters: The function treats `firstYearValue`, `secondYearValue`, and `thirdYearValue` as optional parameters with default values of null. However, in the calculation of `averageHistoricValue`, the code checks whether these values are null and replaces them with 0 if they are. This inconsistency may lead to unexpected results if the default values are not explicitly provided.

I know we discussed original approaches, but looking further there may be possible incorrect variance calculations. 
 The code calculates the percentage change as `(currentValue - averageHistoricValue) / averageHistoricValue`. While this calculation gives the relative change between the current value and the average historical value, it does not represent the variance in the conventional statistical sense. Variance is typically calculated as the average of the squared differences between each value and the mean. The current implementation does not follow this standard variance calculation.

Potential division by zero: If `averageHistoricValue` is zero, the calculation of `percentageChange` will result in a division by zero error. This situation should be handled to avoid potential crashes or unexpected behavior.

Inconsistent variance trigger condition: The code checks if the variance is triggered by comparing `variancePercentage` - `percentageChangeAbs` with `zero`. However, the condition checks if the difference is less than zero, which is inconsistent with the intention of triggering the variance if the difference is less than or equal to zero. The condition should be adjusted accordingly.

Limited flexibility with the number of years: The current implementation assumes a fixed number of historical values (three years). If you want to calculate the variance based on a different number of years, the code would need to be modified. It would be more flexible to accept an array of historical values as a parameter instead of individual parameters for each year.

Reply:
Decision: TBD later.